engine/src/render/shadow_atlas/plugin.cpp

```diff
@@ -0,0 +1,111 @@
+#include <cubos/core/io/window.hpp>
+#include <cubos/core/reflection/external/primitives.hpp>
+
+#include <cubos/engine/render/shadow_atlas/plugin.hpp>
+#include <cubos/engine/render/shadow_atlas/shadow_atlas.hpp>
+#include <cubos/engine/render/shadows/plugin.hpp>
+#include <cubos/engine/render/shadows/spot_caster.hpp>
+#include <cubos/engine/window/plugin.hpp>
+
+using cubos::core::gl::Texture2DDesc;
+using cubos::core::gl::TextureFormat;
+using cubos::core::gl::Usage;
+using cubos::core::io::Window;
+
+CUBOS_DEFINE_TAG(cubos::engine::createShadowAtlasTag);
+CUBOS_DEFINE_TAG(cubos::engine::reserveShadowCastersTag);
+CUBOS_DEFINE_TAG(cubos::engine::drawToShadowAtlasTag);
+
+void cubos::engine::shadowAtlasPlugin(Cubos& cubos)
+{
+    cubos.depends(shadowsPlugin);
+    cubos.depends(windowPlugin);
+
+    cubos.resource<ShadowAtlas>();
+
+    cubos.tag(createShadowAtlasTag);
+    cubos.tag(reserveShadowCastersTag).after(createShadowAtlasTag);
+    cubos.tag(drawToShadowAtlasTag).after(reserveShadowCastersTag);
+
+    cubos.system("create ShadowAtlas").tagged(createShadowAtlasTag).call([](const Window& window, ShadowAtlas& atlas) {
+        if (atlas.size != atlas.configSize)
+        {
+            atlas.size = atlas.configSize;
+
+            // Prepare common texture description.
+            Texture2DDesc desc{};
+            desc.width = atlas.size.x;
+            desc.height = atlas.size.y;
+            desc.usage = Usage::Dynamic;
+
+            auto& rd = window->renderDevice();
+
+            // Create shadow atlas texture.
+            desc.format = TextureFormat::Depth32;
+            atlas.atlas = rd.createTexture2D(desc);
+
+            CUBOS_INFO("Resized ShadowAtlas to {}x{}", atlas.size.x, atlas.size.y);
+        }
+
+        // New frame, hint that the shadow atlas texture needs to be cleared.
+        atlas.cleared = false;
+    });
+
+    cubos.system("reserve space for shadow casters")
```

Comment:
For now this algorithm works and LGTM! It would be cool to improve it in the future though, to allow, for example, assigning a priority to increase the resolution of given lights. If you can, open an issue for this. This could be used for getting extra resolution on lights closer to the camera, and lower on those farther away.

Reply:
Will open after this is merged!